### PR TITLE
feat: Allow passing props to layout component

### PR DESCRIPTION
**Pull request by yaquawa**
Add the ability to pass props to layout component.
Solves https://github.com/ktsn/vue-cli-plugin-auto-routing/issues/61 with no breaking changes.

For some reason, the jest snapshot didn't match when I run the test before I have added any changes, seems like the output html get extra new lines and white spaces, but the structure is the same as before, so I updated the snapshot.

### CODE REVIEW

**Comment by yaquawa** [comment]
Hi @ktsn, could you tell me if you are going to merge this PR?
I'm totally fine even if you're going to reject this, I just want to know if it's Yes or No. Thanks!

File: .gitignore [hunk 1]
```diff
@@ -4,3 +4,4 @@ typings/
 /lib/
 /dist/
 /.tmp/
+.idea
```
Comment (ktsn) [inline .gitignore:7]: Can you revert this change? This kind of file should be put your local gitignore.

File: package.json [hunk 2]
```diff
@@ -56,14 +56,14 @@
     }
   },
   "devDependencies": {
-    "@types/jest": "^26.0.19",
+    "@types/jest": "^26.0.22",
```
Comment (ktsn) [inline package.json:59]: Can you revert all package updates?

File: src/index.ts [hunk 3]
```diff
@@ -8,13 +8,32 @@ import {
 } from 'vue'
 import { RouteRecord, RouteLocationNormalized } from 'vue-router'
 
+type Layout = { name: string; props: Record<string, unknown> }
+
+function normalizeLayout(layout: any): Layout {
```
Comment (ktsn) [inline src/index.ts:13]:
```suggestion
function normalizeLayout(layout: string | { name: string, props?: Record<string, unknown> }): Layout {
```

File: src/index.ts [hunk 4]
```diff
@@ -8,13 +8,32 @@ import {
...
+  }
+
+  if (layout && typeof layout === 'object' && 'name' in layout) {
+    return {
+      name: layout.name,
+      props: layout?.props || {},
+    }
+  }
+
+  throw new Error(`Incorrect value for the 'layout' property.`)
```
Comment (ktsn) [inline src/index.ts:28]:
```suggestion
  throw new Error(`[vue-router-layout] Incorrect value for the 'layout' property.`)
```

File: src/index.ts [hunk 5]
```diff
@@ -107,15 +131,15 @@ export function createRouterLayout(
     async beforeRouteEnter(to, _from, next) {
       await loadAsyncComponents(to)
 
-      const name = resolveLayoutName(to.matched)
-      const layoutComp = name
-        ? normalizeEsModuleComponent(await resolve(name))
+      const layout = resolveLayout(to.matched)
+      const layoutComp = layout?.name
```
Comment (ktsn) [inline src/index.ts:135]:
```suggestion
      const layoutComp = layout
```

File: src/index.ts [hunk 6]
```diff
@@ -154,6 +182,6 @@ export default {
 
 declare module '@vue/runtime-core' {
   interface ComponentCustomOptions {
-    layout?: string
+    layout?: string | Layout
```
Comment (ktsn) [inline src/index.ts:185]:
```suggestion
    layout?: string | { name: string, props?: Record<string, unknown> }
```

**Review by ktsn — CHANGES_REQUESTED** [review]

File: src/index.ts (hunk 5 above)
Comment (yaquawa) [reply src/index.ts:135]: I think this is a wrong suggestion, the `layout` is a normalized Layout object now for example `{name:string, props:{...}}`

Comment (ktsn) [reply src/index.ts:135]: The point of the suggestion is we only need null-check for `layout` but not `layout.name`

Comment (yaquawa) [reply src/index.ts:135]: but then you are going to use `layout.name` right?

see:
https://github.com/yaquawa/vue-router-layout/blob/ff0931c6408412efb9ad18ebd99460c7d41dfd6d/src/index.ts#L142

File: src/index.ts (hunk 3 above)
Comment (yaquawa) [reply src/index.ts:13]: I think it makes sense that the argument expected to be `any` as this is a "normalize" function, anything that couldn't be normalized will just simply throw an error.

**Comment by yaquawa** [comment]
@ktsn Thanks for your review, I've done and did a rebase for it.

**Comment by yaquawa** [comment]
Hi @ktsn , I'm sorry to bother you... but I haven't heard from you for a long time.
I'm using your library for my current project, hopefully this can be merged sooner.
Let me know if I have missed something else.
Thanks🙏🏼

**Review by ktsn — APPROVED** [review]
